src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -99,18 +100,42 @@ public static SolutionStateChecksums Deserialize(ObjectReader reader)
             result[FrozenSourceGeneratedDocumentText] = await SerializableSourceText.FromTextDocumentStateAsync(state.FrozenSourceGeneratedDocumentState, cancellationToken).ConfigureAwait(false);
         }
 
-        foreach (var (_, projectState) in state.ProjectStates)
+        ChecksumCollection.Find(state.AnalyzerReferences, AnalyzerReferences, searchingChecksumsLeft, result, cancellationToken);
+
+        // Before doing a depth-first-search *into* each project, first run across all the project at their top level.
+        // This ensures that when we are trying to sync the projects referenced by a SolutionStateChecksums' instance
+        // that we don't unnecessarily walk all documents looking just for those.
+
+        foreach (var (projectId, projectState) in state.ProjectStates)
```

Comment:
>state.ProjectStates [](http://example.com/codeflow?start=50&length=19)

I wonder if the hintProject case would be better by just creating an ImmutableDictionary with the one entry in it and enumerating over it in the two loops here, rather than iterating over all the project states

Reply:
this changes in a followup pr.  i use the hint project to just index directly into state.ProjectStates.